FeatureRequest: Better exception logging in case of feed processing errors

In case a `station_status` feed (e.g. [this one](https://gbfs.nextbike.net/maps/gbfs/v2/nextbike_ds/de/station_status.json)) does not provide `vehicle_types_available`, a (caught) NPE is reported (see below).

I'm a bit in conflict, if lamassu should handle such issues and work around erroneous feeds to have some cleaner error messages/warnings in the logs. On the other hand, starting to fix and work around might be the road to hell, as there will always be issues which should be fixed at the root and not worked around...

Anyway, exception logs would be much more helpful, if they provided the corrupted feed's system_id.

```
 {"serviceContext":{"service":"lamassu"},"message":"Caught exception in ForkJoinPool\njava.lang.NullPointerException: null\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)\n\tat java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)\n\tat java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)\n\tat java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)\n\tat java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:564)\n\tat java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:591)\n\tat java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:689)\n\tat java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)\n\tat java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)\n\tat java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)\n\tat java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:765)\n\tat org.entur.gbfs.GbfsSubscriptionManager.lambda$update$0(GbfsSubscriptionManager.java:68)\n\tat java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)\nCaused by: java.lang.NullPointerException: Cannot invoke \"java.util.List.stream()\" because the return value of \"org.entur.lamassu.model.entities.Station.getVehicleTypesAvailable()\" is null\n\tat org.entur.lamassu.util.SpatialIndexIdUtil.createStationSpatialIndexId(SpatialIndexIdUtil.java:43)\n\tat org.entur.lamassu.leader.entityupdater.StationsUpdater.lambda$addOrUpdateStations$4(StationsUpdater.java:180)\n\tat java.base/java.util.HashMap.forEach(HashMap.java:1421)\n\tat org.entur.lamassu.leader.entityupdater.StationsUpdater.addOrUpdateStations(StationsUpdater.java:179)\n\tat org.entur.lamassu.leader.entityupdater.EntityCachesUpdater.updateEntityCaches(EntityCachesUpdater.java:55)\n\tat org.entur.lamassu.leader.FeedUpdater.receiveUpdate(FeedUpdater.java:164)\n\tat org.entur.lamassu.leader.FeedUpdater.lambda$createSubscription$1(FeedUpdater.java:126)\n\tat org.entur.gbfs.GbfsSubscription.update(GbfsSubscription.java:104)\n\tat java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)\n\tat java.base/java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(ConcurrentHashMap.java:3612)\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)\n\tat java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)\n\tat java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)\n\t... 5 common frames omitted\n","reportLocation":{"filePath":"org.entur.lamassu.leader.FeedUpdater","lineNumber":"95","functionName":"lambda$start$0"},"severity":"WARNING"}
 ```

--- a/pom.xml
+++ b/pom.xml
@@ -27,7 +27,7 @@
         <okhttp.version>4.12.0</okhttp.version>
         <kryo.version>5.6.0</kryo.version>
         <jackson.version>2.17.2</jackson.version>
-        <gbfs-loader-java.version>4.0.10</gbfs-loader-java.version>
+        <gbfs-loader-java.version>4.0.11</gbfs-loader-java.version>
         <gbfs-validator-java.version>2.0.7</gbfs-validator-java.version>
         <gbfs-mapper-java.version>2.0.8</gbfs-mapper-java.version>
         <prettier-java.version>2.1.0</prettier-java.version>

--- a/src/main/java/org/entur/lamassu/leader/FeedUpdater.java
+++ b/src/main/java/org/entur/lamassu/leader/FeedUpdater.java
@@ -136,6 +136,8 @@ public class FeedUpdater {
       enableValidation
     );
 
+    var interceptor = new LoggingSubscriptionUpdateInterceptor(feedProvider);
+
     String id;
 
     if (feedProvider.getVersion() != null && feedProvider.getVersion().startsWith("3")) {
@@ -150,7 +152,8 @@ public class FeedUpdater {
             );
             receiveV3Update(feedProvider, gbfsV3Delivery);
             cacheReady.set(true);
-          }
+          },
+          interceptor
         );
     } else {
       id =
@@ -164,7 +167,8 @@ public class FeedUpdater {
               GbfsFeedVersionMappers.map(gbfsV2Delivery, feedProvider.getLanguage())
             );
             cacheReady.set(true);
-          }
+          },
+          interceptor
         );
     }
 

--- a/src/main/java/org/entur/lamassu/leader/LoggingSubscriptionUpdateInterceptor.java
+++ b/src/main/java/org/entur/lamassu/leader/LoggingSubscriptionUpdateInterceptor.java
@@ -22,21 +22,22 @@ import org.entur.gbfs.SubscriptionUpdateInterceptor;
 import org.entur.lamassu.model.provider.FeedProvider;
 import org.slf4j.MDC;
 
-public class LoggingSubscriptionUpdateInterceptor implements SubscriptionUpdateInterceptor {
-    private final FeedProvider feedProvider;
+public class LoggingSubscriptionUpdateInterceptor
+  implements SubscriptionUpdateInterceptor {
 
-    public LoggingSubscriptionUpdateInterceptor(
-            FeedProvider feedProvider
-    ) {
-        this.feedProvider = feedProvider;
-    }
-    @Override
-    public void beforeUpdate() {
-        MDC.put("systemId", feedProvider.getSystemId());
-    }
+  private final FeedProvider feedProvider;
 
-    @Override
-    public void afterUpdate() {
-        MDC.remove("systemId");
-    }
+  public LoggingSubscriptionUpdateInterceptor(FeedProvider feedProvider) {
+    this.feedProvider = feedProvider;
+  }
+
+  @Override
+  public void beforeUpdate() {
+    MDC.put("systemId", feedProvider.getSystemId());
+  }
+
+  @Override
+  public void afterUpdate() {
+    MDC.remove("systemId");
+  }
 }

--- a/src/main/java/org/entur/lamassu/leader/LoggingSubscriptionUpdateInterceptor.java
+++ b/src/main/java/org/entur/lamassu/leader/LoggingSubscriptionUpdateInterceptor.java
@@ -1,0 +1,42 @@
+/*
+ *
+ *
+ *  * Licensed under the EUPL, Version 1.2 or – as soon they will be approved by
+ *  * the European Commission - subsequent versions of the EUPL (the "Licence");
+ *  * You may not use this work except in compliance with the Licence.
+ *  * You may obtain a copy of the Licence at:
+ *  *
+ *  *   https://joinup.ec.europa.eu/software/page/eupl
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the Licence is distributed on an "AS IS" basis,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the Licence for the specific language governing permissions and
+ *  * limitations under the Licence.
+ *
+ */
+
+package org.entur.lamassu.leader;
+
+import org.entur.gbfs.SubscriptionUpdateInterceptor;
+import org.entur.lamassu.model.provider.FeedProvider;
+import org.slf4j.MDC;
+
+public class LoggingSubscriptionUpdateInterceptor implements SubscriptionUpdateInterceptor {
+    private final FeedProvider feedProvider;
+
+    public LoggingSubscriptionUpdateInterceptor(
+            FeedProvider feedProvider
+    ) {
+        this.feedProvider = feedProvider;
+    }
+    @Override
+    public void beforeUpdate() {
+        MDC.put("systemId", feedProvider.getSystemId());
+    }
+
+    @Override
+    public void afterUpdate() {
+        MDC.remove("systemId");
+    }
+}

--- a/src/main/resources/logback.xml
+++ b/src/main/resources/logback.xml
@@ -30,6 +30,9 @@
                                     "serviceContext": {
                                         "service": "lamassu"
                                     },
+                                    "mdc": {
+                                        "systemId": "%X{systemId}"
+                                    },
                                     "message": "%message\n%ex{full}",
                                     "severity": "%level",
                                     "reportLocation": {

